ruby/ql/lib/codeql/ruby/security/ImproperMemoizationQuery.qll

```diff
@@ -0,0 +1,87 @@
+/**
+ * Provides predicates for reasoning about improper memoization methods.
+ */
+
+private import ruby
+private import codeql.ruby.DataFlow
+private import codeql.ruby.dataflow.internal.DataFlowDispatch
+
+/**
+ * A `||=` statement that memoizes a value by storing it in an instance variable.
+ */
+private class MemoStmt extends AssignLogicalOrExpr {
+  private InstanceVariableAccess instanceVariable;
+
+  MemoStmt() { instanceVariable.getParent*() = this.getLeftOperand() }
+
+  /**
+   * Gets the variable access on the LHS of this statement.
+   * This is the `@a` in `@a ||= e`.
+   */
+  VariableAccess getVariableAccess() { result = instanceVariable }
+}
+
+/**
+ * A `||=` statement that stores a value at a particular location in a Hash,
+ * which itself is stored in an instance variable.
+ * For example:
+ * ```rb
+ * @a[key] ||= e
+ * ```
+ */
+private class HashMemoStmt extends MemoStmt {
+  HashMemoStmt() {
+    exists(ElementReference e |
+      e = this.getLeftOperand() and this.getVariableAccess().getParent+() = e
+    )
+  }
+}
+
+/**
+ * A method that may be performing memoization.
+ */
+private class MemoCandidate extends Method {
+  MemoCandidate() { exists(MemoStmt m | m.getEnclosingMethod() = this) }
```

Comment:
Nitpick – I think we generally prefer to write it this way. At least, that's what Tom tells me :)
```suggestion
  MemoCandidate() { this = any(MemoStmt m).getEnclosingMethod() }
```